internal/db/service/database.go

```diff
@@ -0,0 +1,230 @@
+		}
+		return nil, fmt.Errorf("failed to create metadata tables: %w", err)
+	}
+
+	service := &DatabaseService{
+		metadataDB: metadataDB,
+		groupDBs:   make(map[string]*sql.DB),
+		dbDir:      dbDir,
+		logger:     logger,
+	}
+
+	return service, nil
+}
+
+func (s *DatabaseService) Close() error {
+	s.groupDBMutex.Lock()
+	defer s.groupDBMutex.Unlock()
+
+	// Close all group databases
+	for groupID, db := range s.groupDBs {
+		if err := db.Close(); err != nil {
+			s.logger.Error("Failed to close group database", "group_id", groupID, "error", err)
+		}
+	}
+
+	// Close metadata database
+	if err := s.metadataDB.Close(); err != nil {
+		s.logger.Error("Failed to close metadata database", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+func createMetadataTables(ctx context.Context, db *sql.DB) error {
+	queries := []string{
+		`CREATE TABLE IF NOT EXISTS groups (
+			group_id TEXT PRIMARY KEY,
+			group_name TEXT NOT NULL
+		) STRICT`,
+		`CREATE TABLE IF NOT EXISTS servers (
+			server_id TEXT NOT NULL,
+			output_channel INTEGER NOT NULL,
+			group_id TEXT NOT NULL REFERENCES groups(group_id),
+			permission_level INTEGER NOT NULL,
+			PRIMARY KEY (server_id, group_id)
+		) STRICT`,
+	}
+
+	for _, query := range queries {
+		if _, err := db.ExecContext(ctx, query); err != nil {
+			return fmt.Errorf("failed to execute query %q: %w", query, err)
+		}
+	}
+
+	return nil
+}
+
+// configureConnection applies Rails-inspired PRAGMA settings for optimal SQLite performance
+func configureConnection(ctx context.Context, db *sql.DB, logger *slog.Logger) error {
+	pragmas := []string{
+		"PRAGMA foreign_keys=ON",              // Enable foreign key constraints for data integrity
+		"PRAGMA journal_mode=WAL",             // Enable WAL mode for better concurrency
+		"PRAGMA synchronous=NORMAL",           // Balance between safety and performance
+		"PRAGMA mmap_size=134217728",          // 128MB memory mapping
+		"PRAGMA journal_size_limit=67108864", // 64MB WAL file limit (triggers auto-checkpoint)
+		"PRAGMA cache_size=2000",              // 2000 pages cache (~8MB with 4KB pages)
+	}
+
+	for _, pragma := range pragmas {
+		// Use QueryContext for PRAGMA statements as they return results
+		rows, err := db.QueryContext(ctx, pragma)
+		if err != nil {
+			// Log the error but continue with other pragmas
+			logger.Warn("Failed to execute PRAGMA", "pragma", pragma, "error", err)
+			
+			// Only fail on critical pragmas
+			if pragma == "PRAGMA foreign_keys=ON" || pragma == "PRAGMA journal_mode=WAL" {
+				return fmt.Errorf("failed to execute critical pragma %q: %w", pragma, err)
+			}
+		} else {
+			// Close the result set immediately
+			if err := rows.Close(); err != nil {
+				logger.Warn("Failed to close rows", "error", err)
+			}
+			logger.Debug("Applied PRAGMA successfully", "pragma", pragma)
+		}
+	}
+
+	return nil
+}
+
+func (s *DatabaseService) getOrCreateGroupDB(ctx context.Context, groupID string) (*sql.DB, error) {
```

Comment:
I don't think I like this. We should be very explicit with regards to when we create a group, since we want the users to be intentional about making/joining groups. We should have this be a get method, and have a separate create method that we call explicitly in the bot/backend